### ISSUE TITLE
generated simple dataset model history

### ISSUE THREAD

**Issue by karino2** [karino2/tegashiki#3]
Original dataset seems too difficult (too small size for it's complexity).
I generate far easier dataset that contains

- 58k one symbol dataset (alphabet and number only)
- 60k NUMBER NUMCHAR (eg. 3a, 2x, 15)
- 60k CHAR_NUMCHAR (eg. X_i, x_2)
- 60k CHAR^NUMCHAR (eg. X^i, y^2)
- add small randomization for each symbol placement, simple data augmentation

from subtask symbol training dataset.
Validation set is almost the same ratio by subtask validation set dataset.

https://github.com/karino2/tegashiki/blob/master/tegashiki_mathexp_generate.ipynb

**Comment by karino2** [comment]
### expgen_rnn_small_dropout05

Model is the same as padstroke_small_rnn_small_dropout05 in #1 .
Feature extractor with GRU encoder-decoder-attention model.

acc 0.956

![expgen_rnn_small_dropout05](https://user-images.githubusercontent.com/1468643/59395842-e8c56a80-8dc0-11e9-9861-daea6a561c85.png)

Impressive score.
It seems generating data from symbol data seems better strategy for current stage.

It's not clear why the score is better than single symbol prediction.
But this dataset contains only alphabet (include common math symbol) and number, so it might be easier to distinguish than one symbol task.

**Comment by karino2** [comment]
Above model is very nice, but it was hard to transform to TensorFlow Lite because RNN is only experimental stage (their dynamic_rnn function generate grap that is hard to train in TPU (dynamic shape even though I supply all shape)).

So I explorer CNN based encoder-decoder model instead.

**Comment by karino2** [comment]
### convdec

Feature extractor create list of stroke features.
Decoder conv1d-ed teacher force input.
Add attention for this output and stroke features.
Add absolute position to decoder input embed.

conv1d filter size is 3, kernel size is 8.

acc: 0.2

![convdec](https://user-images.githubusercontent.com/1468643/59894418-881adb00-941b-11e9-8a83-742308b7b0ec.png)

Out of question...

**Comment by karino2** [comment]
### convf5k128

Above result experiment seems training loss itself is also not decrease enough. 
So just add more parameter in conv1d. kernel size 5. filter size 128.

acc: 0.237

![convdec_f5k128](https://user-images.githubusercontent.com/1468643/59894540-eba50880-941b-11e9-9958-628bf0114d80.png)

Still score is out of question. (though graph shape is much nicer...)

**Comment by karino2** [comment]
### convdec_f5k128_storkeposenc

Add absolute position to stroke feature too.

acc 0.275

![convdec_f5k128_storkeposenc](https://user-images.githubusercontent.com/1468643/59894654-5e15e880-941c-11e9-93f8-88a8f4ca543f.png)

Out of question.

**Comment by karino2** [comment]
### convencdec

Add conv1d to encoder side too.

acc: 0.233

![convencdec](https://user-images.githubusercontent.com/1468643/59894737-ae8d4600-941c-11e9-9108-4e4401be11e6.png)

no improments.

**Comment by karino2** [comment]
### convencdec_posfc

Add embedding layer to abosolute pos before adding up.

acc 0.86

![convencdec_posfc](https://user-images.githubusercontent.com/1468643/59894787-de3c4e00-941c-11e9-8e5d-9a1eaa7a5906.png)

Yes! getting better! I mis-read ConvS2S paper. We need embedding layer.

Score is lower than GRU based model. But this is worth trying score.
Let's covert to TF Lite.

**Comment by karino2** [comment]
### convencdec_myembed2 

Embedding layer with generated tensor (not input) cause TOCO converter to fail. (positional encoding needs this).

It seems tf.gather and dynamically generated tensor cause TOCO converter failure (?).

So I create my own embedding layer.
Create one_hot vector and matmul to -1 to 1 uniform initialized weight matrix.

acc: 0.73

![convencdec_myembed2](https://user-images.githubusercontent.com/1468643/59895027-de891900-941d-11e9-83db-ef709d2cb5ae.png)

Getting score worse. But still working and can convert to TF Lite model.

**Comment by karino2** [comment]
### convencdec_myembed_smallinit

Keras Embedding layer seems weight matrix initialized by -0.05 to 0.05 uniform distribution.
I setup the same initialization.

acc: 0.84

![convencdec_myembed_smallinit](https://user-images.githubusercontent.com/1468643/59895093-2f007680-941e-11e9-8834-45ec4427a770.png)

Now score becomes almost identical to keras Embedding layer (though regularization seems a little different).

Anyway, we have working model that can convert to TF Lite at last!

**Comment by karino2** [comment]
### convencdec_fixfutureleak

Previous model was bug that future input information is leaked via layer normalization.
So just drop this layer and train.

acc: 0.75

![convencdec_fixfutureleak](https://user-images.githubusercontent.com/1468643/59964341-6a906300-953a-11e9-9bb0-3fe990da69b3.png)

Getting worse, but better than random. So this is enough to check in real device.
